esp-rtos/src/esp_radio/timer_queue.rs

```diff
@@ -300,17 +309,33 @@ impl TimerImplementation for Timer {
         TIMER_QUEUE.inner.with(|q| {
             let timer = unsafe { Box::from_raw(timer.cast::<Timer>().as_ptr()) };
 
-            // There are two cases:
-            // - We can remove the timer from the queue - we can drop it.
-            // - We can't remove the timer from the queue. There are the following cases:
-            //   - The timer isn't in the queue. We can drop it.
-            //   - The timer is in the queue and the queue is being processed. We need to mark it to
-            //     be dropped by the timer queue.
-            if q.dequeue(&timer) {
-                core::mem::drop(timer);
-            } else {
+            // There are multiple cases:
+            // - the queue is currently processing - mem-forget the timer and mark it to be dropped
+            //   later in during processing.
+            //
+            // - the queue is currently not processing
+            //      - we can dequeue the timer -> drop it
+            //      - we can't dequeue the timer (i.e. it's not in the queue) -> drop it
+            if q.processing {
                 timer.properties(q).drop = true;
-                core::mem::forget(timer);
+
+                if timer.properties(q).enqueued {
+                    // it's queued so process will make sure to drop it
+                    debug!("schedule timer for dropping after processing the queue");
+                    q.scheduled_for_drop.push(timer);
+                } else {
+                    // drop it since process won't do it for us
+                    debug!("dropping not enqueued timer while processing the queue");
+                    core::mem::drop(timer);
+                }
+            } else {
+                if q.dequeue(&timer) {
+                    debug!("dropping dequeued timer");
+                    core::mem::drop(timer);
+                } else {
+                    debug!("dropping timer not in queue, queue is not processing");
+                    core::mem::drop(timer);
+                }
```

Comment:
I think all we should do is add the timer to the delete list, and wake up the timer task if next_wakeup is infinite. That small delay should be okay, and if the task never gets scheduled, the timers don't work anyway.